# Data Preprocessing - Sample 1

#### Hi, I'm Bryan Bedural. This notebook is a demonstration of some tasks in preprocessing data.

##  <span style="color:red">Part 1:</span>
#### Using data found in sales.csv, I performed the following tasks:

## <span style="color:red">Part1 - 1.</span> Standardize the Price column

In [2]:
import pandas as pd
import numpy as np

data = pd.read_csv("sales.csv")
data

,Transaction_date,Product,Price,Payment_Type,Name,City,State,Country,Account_Created,Last_Login,Latitude,Longitude
0,1/2/2009 6:17,Product1,1200,Mastercard,carolina,Basildon,England,United Kingdom,1/2/2009 6:00,1/2/2009 6:08,51.500000,-1.116667
1,1/2/2009 4:53,Product1,1200,Visa,Betina,Parkville,MO,United States,1/2/2009 4:42,1/2/2009 7:49,39.195000,-94.681940
2,1/2/2009 13:08,Product1,1200,Mastercard,Federica e Andrea,Astoria,OR,United States,1/1/2009 16:21,1/3/2009 12:32,46.188060,-123.830000
3,1/3/2009 14:44,Product1,1200,Visa,Gouya,Echuca,Victoria,Australia,9/25/2005 21:13,1/3/2009 14:22,-36.133333,144.750000
4,1/4/2009 12:56,Product2,3600,Visa,Gerd W,Cahaba Heights,AL,United States,11/15/2008 15:47,1/4/2009 12:45,33.520560,-86.802500
...,...,...,...,...,...,...,...,...,...,...,...,...
995,1/22/2009 14:25,Product1,1200,Visa,Hans-Joerg,Belfast,Northern Ireland,United Kingdom,11/10/2008 12:15,3/1/2009 3:37,54.583333,-5.933333
996,1/28/2009 5:36,Product2,3600,Visa,Christiane,Black River,Black River,Mauritius,1/9/2009 8:10,3/1/2009 4:40,-20.360278,57.366111
997,1/1/2009 4:24,Product3,7500,Amex,Pamela,Skaneateles,NY,United States,12/28/2008 17:28,3/1/2009 7:21,42.946940,-76.429440
998,1/8/2009 11:55,Product1,1200,Diners,julie,Haverhill,England,United Kingdom,11/29/2006 13:31,3/1/2009 7:28,52.083333,0.433333


In [20]:
#checking the data type of Price series
#could also use:    print(pd.api.types.infer_dtype(data.Price))
data.Price.dtype

dtype('float64')

In [4]:
#Since the series is of string format, it should be converted to float or int
data.Price = data.Price.astype(float)

ValueError: could not convert string to float: '13,000'

##### As indicated in "ValueError", the data.Price series has the element 13,000 which has a comma so we need to replace it first in order to convert the type of the series into int or float so that the Price data can be manipulated numerically

In [5]:
# searching the index of the element 13,000
print(data.Price[data.Price == "13,000"])

560    13,000
Name: Price, dtype: object


In [6]:
#creating a dictionary to replace 13,000 with 13000
price2 = {"13,000":"13000"}

#replacing the 13,000 in the Price series of the data dataframe
data = data.replace({"Price":price2})

data.Price[560]

'13000'

In [7]:
# converting Price series str to float
data.Price = data.Price.astype(float)
print(pd.api.types.infer_dtype(data.Price))

floating


In [8]:
#inserting a new column for the standardized Price
data.insert(12, "standardized_price", 0)

In [9]:
# computing and inserting the standardized price values to the new column
price_mean = np.mean(data.Price) #mean of Price
price_std = np.std(data.Price) #std dev of Price

data.standardized_price = (data.Price - price_mean) / price_std

In [10]:
data

,Transaction_date,Product,Price,Payment_Type,Name,City,State,Country,Account_Created,Last_Login,Latitude,Longitude,standardized_price
0,1/2/2009 6:17,Product1,1200.0,Mastercard,carolina,Basildon,England,United Kingdom,1/2/2009 6:00,1/2/2009 6:08,51.500000,-1.116667,-0.372712
1,1/2/2009 4:53,Product1,1200.0,Visa,Betina,Parkville,MO,United States,1/2/2009 4:42,1/2/2009 7:49,39.195000,-94.681940,-0.372712
2,1/2/2009 13:08,Product1,1200.0,Mastercard,Federica e Andrea,Astoria,OR,United States,1/1/2009 16:21,1/3/2009 12:32,46.188060,-123.830000,-0.372712
3,1/3/2009 14:44,Product1,1200.0,Visa,Gouya,Echuca,Victoria,Australia,9/25/2005 21:13,1/3/2009 14:22,-36.133333,144.750000,-0.372712
4,1/4/2009 12:56,Product2,3600.0,Visa,Gerd W,Cahaba Heights,AL,United States,11/15/2008 15:47,1/4/2009 12:45,33.520560,-86.802500,1.610806
...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,1/22/2009 14:25,Product1,1200.0,Visa,Hans-Joerg,Belfast,Northern Ireland,United Kingdom,11/10/2008 12:15,3/1/2009 3:37,54.583333,-5.933333,-0.372712
996,1/28/2009 5:36,Product2,3600.0,Visa,Christiane,Black River,Black River,Mauritius,1/9/2009 8:10,3/1/2009 4:40,-20.360278,57.366111,1.610806
997,1/1/2009 4:24,Product3,7500.0,Amex,Pamela,Skaneateles,NY,United States,12/28/2008 17:28,3/1/2009 7:21,42.946940,-76.429440,4.834024
998,1/8/2009 11:55,Product1,1200.0,Diners,julie,Haverhill,England,United Kingdom,11/29/2006 13:31,3/1/2009 7:28,52.083333,0.433333,-0.372712


##  <span style="color:red">Part1 - 2.</span> Identify and remove outliers in the dataset

In [11]:
#Initial characterization of the data
data.describe(include ="all")

,Transaction_date,Product,Price,Payment_Type,Name,City,State,Country,Account_Created,Last_Login,Latitude,Longitude,standardized_price
count,997,997,999.000000,997,1000,996,987,1000,997,993,997.000000,999.000000,9.990000e+02
unique,979,3,NaN,4,738,747,203,56,968,965,NaN,NaN,NaN
top,1/14/2009 12:07,Product1,NaN,Visa,Sarah,London,England,United States,6/3/2008 4:22,2/26/2009 7:32,NaN,NaN,NaN
freq,2,846,NaN,520,11,19,84,466,3,3,NaN,NaN,NaN
mean,NaN,NaN,1650.970971,NaN,NaN,NaN,NaN,NaN,NaN,NaN,38.984398,-41.417494,-3.737417e-16
std,NaN,NaN,1210.577115,NaN,NaN,NaN,NaN,NaN,NaN,NaN,19.499041,67.377526,1.000501e+00
min,NaN,NaN,20.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,-41.465000,-159.485280,-1.347942e+00
25%,NaN,NaN,1200.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,35.782500,-88.207775,-3.727122e-01
50%,NaN,NaN,1200.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,42.309720,-73.733890,-3.727122e-01
75%,NaN,NaN,1200.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,51.050000,4.850000,-3.727122e-01


### Using the 3-sigma method of removing outliers:

In [12]:
m = np.mean(data.Price) #mean of Price
s = np.std(data.Price) #std dev of Price

print ("Price Mean = " + str(m))
print ("Price Stdev = " + str(s))

ll = max(0, m-3*s)
ul = m + 3*s

print("Price Lower Limit = " + str(ll))
print("Price Upper Limit = " + str(ul))

Price Mean = 1650.970970970971
Price Stdev = 1209.9710691360758
Price Lower Limit = 0
Price Upper Limit = 5280.884178379198


# ------------------------------------------------------------------------------
##### Before cleaning the data using the 3-sigma method, I would like to see information on the Prices first

In [15]:
#Inserting a Count column to get the number of Prices in the dataframe, but used a copy of the data
data_temp = data.copy()
data_temp.insert(13, "Count", 0)

#confirming if original data weren't changed (should have no new column named Count)
data

,Transaction_date,Product,Price,Payment_Type,Name,City,State,Country,Account_Created,Last_Login,Latitude,Longitude,standardized_price
0,1/2/2009 6:17,Product1,1200.0,Mastercard,carolina,Basildon,England,United Kingdom,1/2/2009 6:00,1/2/2009 6:08,51.500000,-1.116667,-0.372712
1,1/2/2009 4:53,Product1,1200.0,Visa,Betina,Parkville,MO,United States,1/2/2009 4:42,1/2/2009 7:49,39.195000,-94.681940,-0.372712
2,1/2/2009 13:08,Product1,1200.0,Mastercard,Federica e Andrea,Astoria,OR,United States,1/1/2009 16:21,1/3/2009 12:32,46.188060,-123.830000,-0.372712
3,1/3/2009 14:44,Product1,1200.0,Visa,Gouya,Echuca,Victoria,Australia,9/25/2005 21:13,1/3/2009 14:22,-36.133333,144.750000,-0.372712
4,1/4/2009 12:56,Product2,3600.0,Visa,Gerd W,Cahaba Heights,AL,United States,11/15/2008 15:47,1/4/2009 12:45,33.520560,-86.802500,1.610806
...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,1/22/2009 14:25,Product1,1200.0,Visa,Hans-Joerg,Belfast,Northern Ireland,United Kingdom,11/10/2008 12:15,3/1/2009 3:37,54.583333,-5.933333,-0.372712
996,1/28/2009 5:36,Product2,3600.0,Visa,Christiane,Black River,Black River,Mauritius,1/9/2009 8:10,3/1/2009 4:40,-20.360278,57.366111,1.610806
997,1/1/2009 4:24,Product3,7500.0,Amex,Pamela,Skaneateles,NY,United States,12/28/2008 17:28,3/1/2009 7:21,42.946940,-76.429440,4.834024
998,1/8/2009 11:55,Product1,1200.0,Diners,julie,Haverhill,England,United Kingdom,11/29/2006 13:31,3/1/2009 7:28,52.083333,0.433333,-0.372712


In [23]:
#Assigning values to the Count column - number of times a specific Price element appeared in the Price column

for i in range (len(data_temp['Price'])):
    data_temp.Count[i] = np.count_nonzero(data.Price==data.Price[i])
    
data_temp

C:\Users\bedural\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


,Transaction_date,Product,Price,Payment_Type,Name,City,State,Country,Account_Created,Last_Login,Latitude,Longitude,standardized_price,Count
0,1/2/2009 6:17,Product1,1200.0,Mastercard,carolina,Basildon,England,United Kingdom,1/2/2009 6:00,1/2/2009 6:08,51.500000,-1.116667,-0.372712,838
1,1/2/2009 4:53,Product1,1200.0,Visa,Betina,Parkville,MO,United States,1/2/2009 4:42,1/2/2009 7:49,39.195000,-94.681940,-0.372712,838
2,1/2/2009 13:08,Product1,1200.0,Mastercard,Federica e Andrea,Astoria,OR,United States,1/1/2009 16:21,1/3/2009 12:32,46.188060,-123.830000,-0.372712,838
3,1/3/2009 14:44,Product1,1200.0,Visa,Gouya,Echuca,Victoria,Australia,9/25/2005 21:13,1/3/2009 14:22,-36.133333,144.750000,-0.372712,838
4,1/4/2009 12:56,Product2,3600.0,Visa,Gerd W,Cahaba Heights,AL,United States,11/15/2008 15:47,1/4/2009 12:45,33.520560,-86.802500,1.610806,137
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,1/22/2009 14:25,Product1,1200.0,Visa,Hans-Joerg,Belfast,Northern Ireland,United Kingdom,11/10/2008 12:15,3/1/2009 3:37,54.583333,-5.933333,-0.372712,838
996,1/28/2009 5:36,Product2,3600.0,Visa,Christiane,Black River,Black River,Mauritius,1/9/2009 8:10,3/1/2009 4:40,-20.360278,57.366111,1.610806,137
997,1/1/2009 4:24,Product3,7500.0,Amex,Pamela,Skaneateles,NY,United States,12/28/2008 17:28,3/1/2009 7:21,42.946940,-76.429440,4.834024,15
998,1/8/2009 11:55,Product1,1200.0,Diners,julie,Haverhill,England,United Kingdom,11/29/2006 13:31,3/1/2009 7:28,52.083333,0.433333,-0.372712,838


In [17]:
#Removing duplicate Prices to the data_temp to show the unique prices for the products
#Named another dataframe to hold those unique values

Unique_Prices = data_temp.drop_duplicates(subset='Price', keep='first', inplace=False, ignore_index=False)
print(Unique_Prices[['Product','Price','Count']])

print("\nNumber of unique prices: " + str(np.count_nonzero(Unique_Prices.Price)))

      Product    Price  Count
0    Product1   1200.0    838
4    Product2   3600.0    137
123  Product1   2100.0      1
152  Product3   7500.0     15
296  Product1    800.0      1
358  Product1   8000.0      1
385  Product1   1250.0      1
442  Product1  10800.0      1
479  Product1    250.0      1
493  Product1   1800.0      1
560  Product1  13000.0      1
647  Product1     20.0      1
721  Product1      NaN      0

Number of unique prices: 13


##### With this Unique_Prices dataframe shown, an initial inspection of the Products and Prices would lead to the conclusion that there should only be 3 prices in the data - one for each Product (Product1, Product2, Product3). The other prices are only one-time input errors for the Product1 price.

##### As such, I believe that the best method to clean the data and remove outliers is by  <span style="color:blue">REMOVING INPUT INACCURACIES</span> in the Product1 Price data. This is because when we use the 3-sigma method, since the price's upper limit that was computed earlier was 5280.88 or below the Product3 price, the Product3 will not be included anymore in the data analysis. Product3 will be removed and treated as an outlier even though it is essential in the data.

In [27]:
#Removing inaccuracies in the Price:

data_cleaned_inaccuracies = data[(data.Price == 1200) | (data.Price == 3600) | (data.Price == 7500)]

data_cleaned_inaccuracies.describe(include="all")

,Transaction_date,Product,Price,Payment_Type,Name,City,State,Country,Account_Created,Last_Login,Latitude,Longitude,standardized_price
count,987,987,990.000000,987,990,986,977,990,987,983,987.000000,989.000000,990.000000
unique,969,3,NaN,4,730,742,203,56,962,959,NaN,NaN,NaN
top,1/4/2009 14:11,Product1,NaN,Visa,Sarah,London,England,United States,6/3/2008 4:22,2/26/2009 7:32,NaN,NaN,NaN
freq,2,836,NaN,516,11,19,84,457,3,3,NaN,NaN,NaN
mean,NaN,NaN,1627.575758,NaN,NaN,NaN,NaN,NaN,NaN,NaN,38.991048,-40.926175,-0.019335
std,NaN,NaN,1103.114985,NaN,NaN,NaN,NaN,NaN,NaN,NaN,19.588293,67.460774,0.911687
min,NaN,NaN,1200.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,-41.465000,-159.485280,-0.372712
25%,NaN,NaN,1200.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,35.960560,-87.953060,-0.372712
50%,NaN,NaN,1200.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,42.375000,-73.616667,-0.372712
75%,NaN,NaN,1200.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,51.066667,4.916667,-0.372712


In [26]:
data_cleaned_inaccuracies

,Transaction_date,Product,Price,Payment_Type,Name,City,State,Country,Account_Created,Last_Login,Latitude,Longitude,standardized_price
0,1/2/2009 6:17,Product1,1200.0,Mastercard,carolina,Basildon,England,United Kingdom,1/2/2009 6:00,1/2/2009 6:08,51.500000,-1.116667,-0.372712
1,1/2/2009 4:53,Product1,1200.0,Visa,Betina,Parkville,MO,United States,1/2/2009 4:42,1/2/2009 7:49,39.195000,-94.681940,-0.372712
2,1/2/2009 13:08,Product1,1200.0,Mastercard,Federica e Andrea,Astoria,OR,United States,1/1/2009 16:21,1/3/2009 12:32,46.188060,-123.830000,-0.372712
3,1/3/2009 14:44,Product1,1200.0,Visa,Gouya,Echuca,Victoria,Australia,9/25/2005 21:13,1/3/2009 14:22,-36.133333,144.750000,-0.372712
4,1/4/2009 12:56,Product2,3600.0,Visa,Gerd W,Cahaba Heights,AL,United States,11/15/2008 15:47,1/4/2009 12:45,33.520560,-86.802500,1.610806
...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,1/22/2009 14:25,Product1,1200.0,Visa,Hans-Joerg,Belfast,Northern Ireland,United Kingdom,11/10/2008 12:15,3/1/2009 3:37,54.583333,-5.933333,-0.372712
996,1/28/2009 5:36,Product2,3600.0,Visa,Christiane,Black River,Black River,Mauritius,1/9/2009 8:10,3/1/2009 4:40,-20.360278,57.366111,1.610806
997,1/1/2009 4:24,Product3,7500.0,Amex,Pamela,Skaneateles,NY,United States,12/28/2008 17:28,3/1/2009 7:21,42.946940,-76.429440,4.834024
998,1/8/2009 11:55,Product1,1200.0,Diners,julie,Haverhill,England,United Kingdom,11/29/2006 13:31,3/1/2009 7:28,52.083333,0.433333,-0.372712


##### The above shows the data for the sales of the 3 products with their corresponding true prices (rows with inaccurate prices were removed)
### --------------------------------------------------------------------------------

### For reference purposes, I will still clean the data using the 3-sigma approach and the IQR approach

#### Continuing with the 3-sigma approach of removing outliers:

In [30]:
print ("Price Mean = " + str(m))
print ("Price Stdev = " + str(s))
print("Price Lower Limit = " + str(ll))
print("Price Upper Limit = " + str(ul))

data_cleaned_3sigma = data[(data.Price >= ll) & (data.Price <= ul)]
data_cleaned_3sigma                                                 

Price Mean = 1650.970970970971
Price Stdev = 1209.9710691360758
Price Lower Limit = 0
Price Upper Limit = 5280.884178379198


,Transaction_date,Product,Price,Payment_Type,Name,City,State,Country,Account_Created,Last_Login,Latitude,Longitude,standardized_price
0,1/2/2009 6:17,Product1,1200.0,Mastercard,carolina,Basildon,England,United Kingdom,1/2/2009 6:00,1/2/2009 6:08,51.500000,-1.116667,-0.372712
1,1/2/2009 4:53,Product1,1200.0,Visa,Betina,Parkville,MO,United States,1/2/2009 4:42,1/2/2009 7:49,39.195000,-94.681940,-0.372712
2,1/2/2009 13:08,Product1,1200.0,Mastercard,Federica e Andrea,Astoria,OR,United States,1/1/2009 16:21,1/3/2009 12:32,46.188060,-123.830000,-0.372712
3,1/3/2009 14:44,Product1,1200.0,Visa,Gouya,Echuca,Victoria,Australia,9/25/2005 21:13,1/3/2009 14:22,-36.133333,144.750000,-0.372712
4,1/4/2009 12:56,Product2,3600.0,Visa,Gerd W,Cahaba Heights,AL,United States,11/15/2008 15:47,1/4/2009 12:45,33.520560,-86.802500,1.610806
...,...,...,...,...,...,...,...,...,...,...,...,...,...
994,1/20/2009 10:42,Product2,3600.0,Diners,esther,Huddersfield,NaN,United Kingdom,1/20/2009 9:15,3/1/2009 3:29,53.650000,-1.783333,1.610806
995,1/22/2009 14:25,Product1,1200.0,Visa,Hans-Joerg,Belfast,Northern Ireland,United Kingdom,11/10/2008 12:15,3/1/2009 3:37,54.583333,-5.933333,-0.372712
996,1/28/2009 5:36,Product2,3600.0,Visa,Christiane,Black River,Black River,Mauritius,1/9/2009 8:10,3/1/2009 4:40,-20.360278,57.366111,1.610806
998,1/8/2009 11:55,Product1,1200.0,Diners,julie,Haverhill,England,United Kingdom,11/29/2006 13:31,3/1/2009 7:28,52.083333,0.433333,-0.372712


In [31]:
data_cleaned_3sigma.describe(include="all")

,Transaction_date,Product,Price,Payment_Type,Name,City,State,Country,Account_Created,Last_Login,Latitude,Longitude,standardized_price
count,978,978,981.000000,979,981,977,968,981,978,974,979.000000,980.000000,981.000000
unique,960,2,NaN,4,727,736,202,56,957,954,NaN,NaN,NaN
top,1/5/2009 5:33,Product1,NaN,Visa,Sarah,London,England,United States,1/18/2009 0:00,1/24/2009 14:54,NaN,NaN,NaN
freq,2,842,NaN,511,11,19,84,455,3,2,NaN,NaN,NaN
mean,NaN,NaN,1534.169215,NaN,NaN,NaN,NaN,NaN,NaN,NaN,39.116890,-41.097999,-0.096533
std,NaN,NaN,834.950840,NaN,NaN,NaN,NaN,NaN,NaN,NaN,19.433659,67.592075,0.690059
min,NaN,NaN,20.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,-41.465000,-159.485280,-1.347942
25%,NaN,NaN,1200.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,36.013475,-87.861807,-0.372712
50%,NaN,NaN,1200.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,42.395830,-73.722083,-0.372712
75%,NaN,NaN,1200.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,51.083333,4.916667,-0.372712


##### As seen above, using the 3-sigma limit approach of removing outliers in these data will result in removing a product with the correct pricing (product3). Upon cleaning the data by subjecting them to the upper and lower limits, the minimum Price value is still 20 (which is obviously an error because as stated in the Count column of the Unique_Prices dataframe, Product1's Price should be 1200), and the max value is now 3600 (price of Product2), removing the rows for the Product3 (priced at 7500). The spread of the Price data may be lower (low std), but this results in an incomplete data which could negatively affect the analysis.

### Now let's try the Inter-Quartile Range (IQR) method:

IQR = Q75 - Q25

In [49]:
IQR = np.percentile(data.Price,75) - np.percentile(data.Price,25)

print("75th percentile of Price data column = " + str(np.percentile(data.Price,75)))
print("25th percentile of Price data column = " + str(np.percentile(data.Price,25)))
print("\n\b IQR = " + str(IQR))

ul2 = np.percentile(data.Price,75) + 1.5*IQR
ll2 = max(0, np.percentile(data.Price,25)-1.5*IQR)

print("\nUpper limit 2 = " + str(ul2))
print("Lower limit 2 = " + str(ll2))

data_cleaned_IQR = data[(data.Price >= ll2) & (data.Price <= ul2)]

print('\033[1m' + "\nCleaned data using the IQR method: ")
data_cleaned_IQR

75th percentile of Price data column = nan
25th percentile of Price data column = nan

 IQR = nan

Upper limit 2 = nan
Lower limit 2 = 0

Cleaned data using the IQR method: 


,Transaction_date,Product,Price,Payment_Type,Name,City,State,Country,Account_Created,Last_Login,Latitude,Longitude,standardized_price


##### This shows that for this dataframe, the IQR approach in removing outliers SHOULD NOT be used. It will only result in removing all of the data since the IQR computed is nan (not a number) which results in considering all of the prices as outliers when they are subjected to the limits.

##  <span style="color:red">Part 2:</span> 
## Handling Missing Values

### Using the data in Handling Missing Values.csv, handle all the missing values and provide an explanation for choosing a specific method (delete, impute or keep: if impute, why mean, median, or mode?) in handling specific missing values

In [357]:
import pandas as pd

part2 = pd.read_csv("Handling_Missing_Values.csv")
print("Number of rows: "+str(part2['ID'].count()))
part2

Number of rows: 20


,ID,Sex,Age,Income,Employed,Children,Buy_Car
0,1,Male,25,25146.0,Single,0.0,No
1,2,Male,30,26939.0,Married,2.0,Yes
2,3,Male,27,26693.0,Married,0.0,No
3,4,Male,28,26666.0,Married,3.0,Yes
4,5,Male,29,25899.0,Married,0.0,No
5,6,Male,28,26462.0,Married,1.0,No
6,7,Female,28,NaN,Married,3.0,Yes
7,8,NaN,30,26037.0,Married,2.0,Yes
8,9,Female,28,26167.0,Married,1.0,Yes
9,10,NaN,28,NaN,Single,NaN,No


In [358]:
part2.describe(include="all")

,ID,Sex,Age,Income,Employed,Children,Buy_Car
count,20.00000,16,20.000000,17.000000,20,18.000000,20
unique,NaN,2,NaN,NaN,2,NaN,2
top,NaN,Male,NaN,NaN,Married,NaN,No
freq,NaN,10,NaN,NaN,15,NaN,11
mean,10.50000,NaN,28.150000,26245.705882,NaN,1.222222,NaN
std,5.91608,NaN,1.308877,583.052395,NaN,1.437136,NaN
min,1.00000,NaN,25.000000,25094.000000,NaN,0.000000,NaN
25%,5.75000,NaN,28.000000,26037.000000,NaN,0.000000,NaN
50%,10.50000,NaN,28.000000,26234.000000,NaN,1.000000,NaN
75%,15.25000,NaN,29.000000,26666.000000,NaN,2.000000,NaN


In [359]:
# summary of missing values
part2.isnull().sum()

ID          0
Sex         4
Age         0
Income      3
Employed    0
Children    2
Buy_Car     0
dtype: int64

In [360]:
#printing the columns that contain null values
null_columns = part2.columns[part2.isnull().any()]
part2[part2.isnull().any(axis=1)][null_columns]

,Sex,Income,Children
6,Female,NaN,3.0
7,NaN,26037.0,2.0
9,NaN,NaN,NaN
11,Male,26195.0,NaN
14,NaN,26234.0,0.0
17,NaN,NaN,0.0


--------------------------------
#### We can see that there are 3 columns that have missing values. It's good that there are no missing values in the Buy_Car column (if there are, I think it's best to delete those rows with missing Buy_Car input because it's the target variable).

#### There are 7 columns, and I think it's best for a specific row or record if 5 of those have no missing values so that there will be enough information for the data. As such, those with 3 or more missing inputs will be deleted

In [361]:
# deleting all rows with less than 5 non-NaN values
part2 = part2.dropna(thresh = 5)
print("Number of rows: " + str(part2['ID'].count()))
part2

Number of rows: 19


,ID,Sex,Age,Income,Employed,Children,Buy_Car
0,1,Male,25,25146.0,Single,0.0,No
1,2,Male,30,26939.0,Married,2.0,Yes
2,3,Male,27,26693.0,Married,0.0,No
3,4,Male,28,26666.0,Married,3.0,Yes
4,5,Male,29,25899.0,Married,0.0,No
5,6,Male,28,26462.0,Married,1.0,No
6,7,Female,28,NaN,Married,3.0,Yes
7,8,NaN,30,26037.0,Married,2.0,Yes
8,9,Female,28,26167.0,Married,1.0,Yes
10,11,Female,28,26905.0,Single,0.0,No


#### Only 1 record was deleted.
-----------------------------------------
#### For the 'Sex' column, I will FILL the missing values with "No Data" because the respondents might have purposefully skipped that. They could have not included one because they could have a different sexual orientation that was not specified in the options. 

#### For the 'Children' column, I will KEEP the missing values since the respondents who missed that could have a purpose for not disclosing the number of their children, like for privacy or safety purposes. Filling those missing values with a "No Data" string will mix a string data type to the numerical column, so it should be avoided if we want to analyze the Children column in the future using functions specialized for numbers (like max, min, mean). 

In [362]:
#fill NaN values of Sex column
part2['Sex'] = part2['Sex'].fillna("No Data") 
part2

C:\Users\bedural\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


,ID,Sex,Age,Income,Employed,Children,Buy_Car
0,1,Male,25,25146.0,Single,0.0,No
1,2,Male,30,26939.0,Married,2.0,Yes
2,3,Male,27,26693.0,Married,0.0,No
3,4,Male,28,26666.0,Married,3.0,Yes
4,5,Male,29,25899.0,Married,0.0,No
5,6,Male,28,26462.0,Married,1.0,No
6,7,Female,28,NaN,Married,3.0,Yes
7,8,No Data,30,26037.0,Married,2.0,Yes
8,9,Female,28,26167.0,Married,1.0,Yes
10,11,Female,28,26905.0,Single,0.0,No


--------------------------------
#### For the 'Income' column, I will handle the missing values using two ways: fill or keep. 

First, if the respondent is <span style="color:blue">Single</span> and will buy a car <span style="color:blue">(Buy_Car = Yes)</span>, I will use the <span style="color:blue">MEAN</span>  to fill up the missing income cell (the distribution of incomes is normal so a mean is preferred) because I think there is a high chance for the respondent to be earning if he or she wants to buy a car even without a partner to rely on for the cost and with the given age range of the respondents (25-30 yrs old is the age range where people have usually stopped asking for gifts or financial support from their parents). 

Second, I will <span style="color:blue">KEEP</span>  the missing income value if the respondent is <span style="color:blue">Married</span>  since this could be meaningful. He or she could be relying on the income of his or her partner (e.g. she's a housewife, or unemployed). I will also keep the missing value if he/she is single and will NOT buy a car because this could indicate that he/she is unemployed.

In [363]:
#Changing the wrong column name: "Employed" should be "Marital Status" based on the column content
part2.rename(columns={'Employed' : 'Marital_Status'}, inplace=True)
part2

C:\Users\bedural\Anaconda3\lib\site-packages\pandas\core\frame.py:5047: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


,ID,Sex,Age,Income,Marital_Status,Children,Buy_Car
0,1,Male,25,25146.0,Single,0.0,No
1,2,Male,30,26939.0,Married,2.0,Yes
2,3,Male,27,26693.0,Married,0.0,No
3,4,Male,28,26666.0,Married,3.0,Yes
4,5,Male,29,25899.0,Married,0.0,No
5,6,Male,28,26462.0,Married,1.0,No
6,7,Female,28,NaN,Married,3.0,Yes
7,8,No Data,30,26037.0,Married,2.0,Yes
8,9,Female,28,26167.0,Married,1.0,Yes
10,11,Female,28,26905.0,Single,0.0,No


In [364]:
# querying (reference only to show the data that satisfy the conditions)
part2_queried = part2[(part2.Marital_Status=='Single') & (part2.Buy_Car=='Yes')]
part2_queried

,ID,Sex,Age,Income,Marital_Status,Children,Buy_Car
11,12,Male,30,26195.0,Single,NaN,Yes
17,18,No Data,28,NaN,Single,0.0,Yes


In [365]:
#getting the mean of incomes
income_mean = part2['Income'].mean() #income_mean = part2.Income.mean()
print("Income_mean = " + str(round(income_mean,2)))

Income_mean = 26245.71


In [366]:
#resetting the index of the dataframe for easier manipulation
part2 = part2.reset_index(drop=True)
#reset_index func adds another column for old indices, default drop=False
#drop=True means dropping the newly generated columns for old indices
#ignore_index=True parameter is not working for dropna function

part2

,ID,Sex,Age,Income,Marital_Status,Children,Buy_Car
0,1,Male,25,25146.0,Single,0.0,No
1,2,Male,30,26939.0,Married,2.0,Yes
2,3,Male,27,26693.0,Married,0.0,No
3,4,Male,28,26666.0,Married,3.0,Yes
4,5,Male,29,25899.0,Married,0.0,No
5,6,Male,28,26462.0,Married,1.0,No
6,7,Female,28,NaN,Married,3.0,Yes
7,8,No Data,30,26037.0,Married,2.0,Yes
8,9,Female,28,26167.0,Married,1.0,Yes
9,11,Female,28,26905.0,Single,0.0,No


##### Notice the indices above, they are now arranged or ordered properly so that we will not have a problem in the loop below

In [367]:
#filling in missing values for the queried data:
for x in range (len(part2['Income'])):
    if ((part2.Marital_Status[x]=='Single') and (part2.Buy_Car[x]=='Yes')):
        if part2.Income[x] > 0:
            part2.Income[x] = part2.Income[x]
        else:
            part2.Income[x]=income_mean
            
part2    

C:\Users\bedural\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
C:\Users\bedural\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


,ID,Sex,Age,Income,Marital_Status,Children,Buy_Car
0,1,Male,25,25146.000000,Single,0.0,No
1,2,Male,30,26939.000000,Married,2.0,Yes
2,3,Male,27,26693.000000,Married,0.0,No
3,4,Male,28,26666.000000,Married,3.0,Yes
4,5,Male,29,25899.000000,Married,0.0,No
5,6,Male,28,26462.000000,Married,1.0,No
6,7,Female,28,NaN,Married,3.0,Yes
7,8,No Data,30,26037.000000,Married,2.0,Yes
8,9,Female,28,26167.000000,Married,1.0,Yes
9,11,Female,28,26905.000000,Single,0.0,No


##### ID 18 now has its income value filled with the Mean. It's the only row that satisfied the conditions (Income  = NaN, Buy_car = Yes, Marital_Status = Single).

##### The other missing income value was kept as is, for reasons stated above.

----------------------------------------------------------